man/Xevi.man
.\" $Xorg: Xevi.man,v 1.3 2000/08/17 19:41:54 cpqbld Exp $
.\"
.\" Copyright (c) 1997  Silicon Graphics, Inc.
.\" 
.\" Permission is hereby granted, free of charge, to any person obtaining a
.\" copy of this software and associated documentation files (the "Software"), 
.\" to deal in the Software without restriction, including without limitation 
.\" the rights to use, copy, modify, merge, publish, distribute, sublicense, 
.\" and/or sell copies of the Software, and to permit persons to whom the 
.\" Software furnished to do so, subject to the following conditions:
.\" 
.\" The above copyright notice and this permission notice shall be included in
.\" all copies or substantial portions of the Software.
.\" 
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
.\" IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
.\" FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL 
.\" SILICON GRAPHICS, INC. BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, 
.\" WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF 
.\" OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE 
.\" SOFTWARE.
.\" 
.\" Except as contained in this notice, the name of Silicon Graphics, Inc. shall not 
.\" be used in advertising or otherwise to promote the sale, use or other 
.\" dealing in this Software without prior written authorization from
.\" Silicon Graphics, Inc.
.\"
.\" $XFree86$
.\"
.TH XEVI 3X11 __xorgversion__ "X FUNCTIONS"
.SH NAME
XeviQueryExtension, XeviQueryVersion, XeviGetVisualInfo - X Extended Visual Information functions
.SH SYNOPSIS
\&#include <X11/extensions/XEVI.h>

Bool
XeviQueryExtension (Display *dpy);
.LP
Bool
XeviQueryVersion (Display *dpy, 
    int *major_version_return,
    int *minor_version_return);
.LP
int
XeviGetVisualInfo (Display *dpy, VisualID *visual,
    int n_visual, ExtendedVisualInfo ** evi_return,
    int * n_info_return);
.fi
.SH DESCRIPTION
The
X11 Extended Visual Information extension (EVI) allows a client to determine
information about core X visuals beyond what the core protocol provides.

The EVI application programming library contains the interfaces described below.
With the exception of
.BR XeviQueryExtension,
if any of these routines are
called with a display that does not support the extension, the
ExtensionErrorHandler (which can be set with
.BR XSetExtensionErrorHandler
and functions the same way as
.BR XSetErrorHandler )
will be called and the function will then return.
.PP
.BR XeviQueryExtension
returns 
.BR True
if the Extended Visual Information extension is available on the given display.
A client must call
.BR XeviQueryExtension
before calling any other EVI
function in order to negotiate a compatible protocol version; otherwise
the client will get undefined behavior (EVI may or may not work).
.PP
.BR XeviQueryVersion
returns
.BR True
if the request succeeded; the values of the major and minor protocol
version supported by the server are returned in
.I major_version_return
and
.I minor_version_return.
.PP
.BR XeviGetVisualInfo
returns a list of ExtendedVisualInfo structures that describe
visual information beyond that supported by the core protocol. This includes 
layer information relevant for systems supporting overlays and/or underlay
planes, and information that allows applications better to determine the level
of hardware support for multiple colormaps. XeviGetVisualInfo returns 
.BR Success 
if successful, or an X error otherwise.
If the argument
.I visual
is NULL, then information for all visuals of all 
screens is returned. Otherwise, it's a pointer to a list of visuals for which
extended visual information is desired.
.I n_visual
is the number of elements in the array visual.
.I evi_return
returns a pointer to a list of ExtendedVisualInfo. When done, the client
should free the list using XFree\.
.I n_info_return
returns the number of elements in the array evi_return.
.LP
The
.BR ExtendedVisualInfo
structure has the following fields:
.in +.5i
.nf
.ta 1.6i
VisualID	\fIcore_visual_id\fR    
int 	\fIscreen\fR	    	
int 	\fIlevel\fR 	    	
unsigned int	\fItransparency_type\fR 
unsigned int	\fItransparency_value\fR 
unsigned int	\fImin_hw_colormaps\fR
unsigned int	\fImax_hw_colormaps\fR
unsigned int	\fInum_colormap_conflicts\fR
VisualID *  	\fIcolormap_conflicts\fR
.fi
.LP
The combination of 
.I core_visual_id
and
.I screen
number uniquely specify the visual being described.
.LP
.I level
returns the level number for the visual, 0 for normal planes, > 0 for
overlays, < 0 for underlays.
.LP
.I
transparency_type
returns the type of transparency supported by the visual. XEVI_TRANSPARENCY_NONE if
there are no transparent pixels, XEVI_TRANSPARENCY_PIXEL if the visual supports
a transparent pixel, XEVI_TRANSPARENCY_MASK if the visual supports transparent
plane(s).
.LP
.I transparency_value
returns the pixel/plane value to set for transparency if transparency_type isn't
XEVI_TRANSPARENCY_NONE.
.LP
.I min_hw_colormaps
and
.I max_hw_colormaps
return the minimum and maximum number of hardware colormaps backing up the
visual. 
.LP
.I num_colormap_conflicts
returns the number of elements in 
.I colormap_conflicts.
This array returns a list of visuals that may cause conflicts in the use of the
hardware colormap. For example, if a 12-bit hardware colormap is overloaded to
support 8-bit colormaps, the corresponding 8-bit visuals would conflict with the
12-bit visuals.

.SH ERRORS
.BR XeviGetVisualInfo
will return
.I BadValue
if passed an illegal visual ID,
.I BadAccess
if the X server does not respond,
.I BadAlloc
if there is a memory allocation failure.


ChangeLog
2005-12-03  Kevin E. Martin  <kem-at-freedesktop-dot-org>

	* configure.ac:
	Update package version number for X11R7 RC3 release.

2005-12-02  Kevin E. Martin  <kem-at-freedesktop-dot-org>

	* configure.ac:
	* src/Makefile.am:
	Add check and cflags for malloc(0) returning NULL.

2005-11-19  Kevin E. Martin  <kem-at-freedesktop-dot-org>

	* xext.pc.in:
	Update pkgconfig files to separate library build-time dependencies
	from application build-time dependencies.

2005-11-09  Kevin E. Martin  <kem-at-freedesktop-dot-org>

	* configure.ac:
	Update package version number for X11R7 RC2 release.

2005-11-01  Kevin E. Martin  <kem-at-freedesktop-dot-org>

	* configure.ac:
	Update pkgcheck dependencies to work with separate build roots.

2005-10-18  Kevin E. Martin  <kem-at-freedesktop-dot-org>

	* configure.ac:
	Update package version number for RC1 release.

2005-10-12  Kevin E. Martin  <kem-at-freedesktop-dot-org>

	* man/Makefile.am:
	Add .shadows.DONE to CLEANFILES to pass distcheck

2005-10-11  Alan Coopersmith  <alan.coopersmith@sun.com>

	* man/Makefile.am: 
	* configure.ac
	Set up cpp pre-processing of man pages
	Add shadow man pages for man pages that document multiple functions.

2005-07-08  Keith Packard  <keithp@keithp.com>

	* .cvsignore:
	* man/.cvsignore:
	* src/.cvsignore:
	* src/Makefile.am:
	Add .cvsignore files
	Switch _la_CFLAGS for AM_CFLAGS to clean up directory

2005-06-28  Alan Coopersmith  <alan.coopersmith@sun.com>

	* man/Makefile.am (man3_MANS): 
	Add man pages for DPMS Extension calls in libXext.   (Converted to
	man page format by Sun doc teams based on DPMS specification docs.)
	
2005-05-22  Alan Coopersmith  <alan.coopersmith@sun.com>

	* man/Makefile.am (man3_MANS): Change man pages to long file names

2005-01-29  Daniel Stone  <daniel@freedesktop.org>

	* XSync.c (wire_to_event):
	Sync from X.Org HEAD -- add explicit cast to prevent warning.

2004-11-22  Carl Worth  <cworth@cworth.org>

	* configure.ac: Remove AC_CONFIG_AUX_DIR (it was in the wrong
	place for automake-1.9, and it was set to the default value
	anyway).

2004-09-19  Carlos Romero  <kidcrash@freedesktop.org>

	* man/Makefile.am:
	Include man pages in EXTRA_DIST.

2004-04-13  Warren Turkal  <wt@midsouth.rr.com>

	* ImUtil.h is not in the public API and is not used in any other parts
	  of the x packages that I can see

2004-04-12  Daniel Stone  <daniel@freedesktop.org>

	* XSync.c:
	Merge from X11R6.7 - do explicit casting.

	* Bump version to 6.4.3 for xlibs 1.0.1.

2004-02-03  Jim Gettys  <jg@freedesktop.org>

	* AUTHORS needed lots of research to get the authors documented

2004-01-15  Daniel Stone  <daniel@fooishbar.org>
	* Tagging release 6.4.2 for the first freedesktop.org clientside lib
	  release.

2004-01-15  Harold L Hunt II <huntharo@msu.edu>
	* Makefile.am: Pass -no-undefined to libtool via LDFLAGS.


src/XEVI.c
/* $Xorg: XEVI.c,v 1.3 2000/08/17 19:45:51 cpqbld Exp $ */
/************************************************************
Copyright (c) 1997 by Silicon Graphics Computer Systems, Inc.
Permission to use, copy, modify, and distribute this
software and its documentation for any purpose and without
fee is hereby granted, provided that the above copyright
notice appear in all copies and that both that copyright
notice and this permission notice appear in supporting
documentation, and that the name of Silicon Graphics not be
used in advertising or publicity pertaining to distribution
of the software without specific prior written permission.
Silicon Graphics makes no representation about the suitability
of this software for any purpose. It is provided "as is"
without any express or implied warranty.
SILICON GRAPHICS DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS
SOFTWARE, INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS FOR A PARTICULAR PURPOSE. IN NO EVENT SHALL SILICON
GRAPHICS BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL
DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE,
DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE
OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION  WITH
THE USE OR PERFORMANCE OF THIS SOFTWARE.
********************************************************/
/* $XFree86$ */
#define NEED_EVENTS
#define NEED_REPLIES
#include <X11/Xlibint.h>
#include <X11/extensions/XEVI.h>
#include <X11/extensions/XEVIstr.h>
#include <X11/extensions/Xext.h>
#include <X11/extensions/extutil.h>
#include <X11/Xutil.h>
static XExtensionInfo *xevi_info;/* needs to move to globals.c */
static /* const */ char *xevi_extension_name = EVINAME;
#define XeviCheckExtension(dpy,i,val) \
  XextCheckExtension (dpy, i, xevi_extension_name, val)
/*****************************************************************************
 *                                                                           *
 *			   private utility routines                          *
 *                                                                           *
 *****************************************************************************/
static /* const */ XExtensionHooks xevi_extension_hooks = {
    NULL,			/* create_gc */
    NULL,			/* copy_gc */
    NULL,			/* flush_gc */
    NULL,			/* free_gc */
    NULL,			/* create_font */
    NULL,			/* free_font */
    NULL,			/* close_display */
    NULL,			/* wire_to_event */
    NULL,			/* event_to_wire */
    NULL,			/* error */
    NULL,			/* error_string */
};
static XEXT_GENERATE_FIND_DISPLAY (find_display, xevi_info,
                                   xevi_extension_name,
                                   &xevi_extension_hooks, 0, NULL)
Bool XeviQueryExtension (Display *dpy)
{
    XExtDisplayInfo *info = find_display (dpy);
    if (XextHasExtension(info)) {
	return True;
    } else {
	return False;
    }
}
Bool XeviQueryVersion(Display *dpy, int *majorVersion, int *minorVersion)
{
    XExtDisplayInfo *info = find_display (dpy);
    xEVIQueryVersionReply rep;
    register xEVIQueryVersionReq *req;
    XeviCheckExtension (dpy, info, False);
    LockDisplay(dpy);
    GetReq(EVIQueryVersion, req);
    req->reqType = info->codes->major_opcode;
    req->xeviReqType = X_EVIQueryVersion;
    if (!_XReply(dpy, (xReply *)&rep, 0, xFalse)) {
	UnlockDisplay(dpy);
	SyncHandle();
	return False;
    }
    *majorVersion = rep.majorVersion;
    *minorVersion = rep.minorVersion;
    UnlockDisplay(dpy);
    SyncHandle();
    return True;
}
static Bool notInList(VisualID32 *visual, int sz_visual, VisualID newVisualid)
{
    while  (sz_visual-- > 0)  {
	if (*visual == newVisualid)
	    return False;
	visual++;
    }
    return True;
}
Status XeviGetVisualInfo(
    register Display *dpy,
    VisualID *visual,
    int n_visual,
    ExtendedVisualInfo **evi_return,
    int *n_info_return)
{
    XExtDisplayInfo *info = find_display (dpy);
    register xEVIGetVisualInfoReq *req;
    xEVIGetVisualInfoReply rep;
    int sz_info, sz_xInfo, sz_conflict, sz_xConflict;
    VisualID32 *temp_conflict, *temp_visual, *xConflictPtr;
    VisualID *conflict;
    xExtendedVisualInfo *temp_xInfo;
    XVisualInfo *vinfo;
    register ExtendedVisualInfo *infoPtr;
    register xExtendedVisualInfo *xInfoPtr;
    register int n_data, visualIndex, vinfoIndex;
    Bool isValid;
    XeviCheckExtension (dpy, info, 0);
    *n_info_return = 0;
    *evi_return = NULL;
    vinfo = XGetVisualInfo(dpy, 0, NULL, &sz_info);
    if (!vinfo) {
	return BadValue;
    }
    if (!n_visual || !visual) {		/* copy the all visual */
    	temp_visual = (VisualID32 *)Xmalloc(sz_VisualID32 * sz_info);
    	n_visual = 0;
        for (vinfoIndex = 0; vinfoIndex < sz_info; vinfoIndex++)
	    if (notInList(temp_visual, n_visual, vinfo[vinfoIndex].visualid))
	        temp_visual[n_visual++] = vinfo[vinfoIndex].visualid;
    }
    else {	/* check if the visual is valid */
        for (visualIndex = 0; visualIndex < n_visual; visualIndex++) {
	    isValid = False;
            for (vinfoIndex = 0; vinfoIndex < sz_info; vinfoIndex++) {
	        if (visual[visualIndex] == vinfo[vinfoIndex].visualid) {
		    isValid = True;
		    break;
	        }
	    }
	    if (!isValid) {
		XFree(vinfo);
	        return BadValue;
	    }
	}
	temp_visual = (VisualID32 *)Xmalloc(sz_VisualID32 * n_visual);
        for (visualIndex = 0; visualIndex < n_visual; visualIndex++)
	    temp_visual[visualIndex] = visual[visualIndex];
    }
    XFree(vinfo);
    LockDisplay(dpy);
    GetReq(EVIGetVisualInfo, req);
    req->reqType = info->codes->major_opcode;
    req->xeviReqType = X_EVIGetVisualInfo;
    req->n_visual = n_visual;
    SetReqLen(req, n_visual, 1);
    Data(dpy, (char *)temp_visual, n_visual * sz_VisualID32);
    if (!_XReply(dpy, (xReply *)&rep, 0, xFalse)) {
	UnlockDisplay(dpy);
	SyncHandle();
	Xfree(temp_visual);
	return BadAccess;
    }
    Xfree(temp_visual);
    sz_info = rep.n_info * sizeof(ExtendedVisualInfo);
    sz_xInfo = rep.n_info * sz_xExtendedVisualInfo;
    sz_conflict = rep.n_conflicts * sizeof(VisualID);
    sz_xConflict = rep.n_conflicts * sz_VisualID32;
    infoPtr = *evi_return = (ExtendedVisualInfo *)Xmalloc(sz_info + sz_conflict);
    xInfoPtr = temp_xInfo = (xExtendedVisualInfo *)Xmalloc(sz_xInfo);
    xConflictPtr = temp_conflict = (VisualID32 *)Xmalloc(sz_xConflict);
    if (!*evi_return || !temp_xInfo || !temp_conflict) {
        _XEatData(dpy, (sz_xInfo + sz_xConflict + 3) & ~3);
	UnlockDisplay(dpy);
	SyncHandle();
	if (evi_return)
	   Xfree(evi_return);
	if (temp_xInfo)
	   Xfree(temp_xInfo);
	if (temp_conflict)
	   Xfree(temp_conflict);
	return BadAlloc;
    }
    _XRead(dpy, (char *)temp_xInfo, sz_xInfo);
    _XRead(dpy, (char *)temp_conflict, sz_xConflict);
    UnlockDisplay(dpy);
    SyncHandle();
    n_data = rep.n_info;
    conflict = (VisualID *)(infoPtr + n_data);
    while (n_data-- > 0) {
	infoPtr->core_visual_id		= xInfoPtr->core_visual_id;
	infoPtr->screen			= xInfoPtr->screen;
	infoPtr->level			= xInfoPtr->level;
	infoPtr->transparency_type	= xInfoPtr->transparency_type;
	infoPtr->transparency_value	= xInfoPtr->transparency_value;
	infoPtr->min_hw_colormaps	= xInfoPtr->min_hw_colormaps;
	infoPtr->max_hw_colormaps	= xInfoPtr->max_hw_colormaps;
	infoPtr->num_colormap_conflicts = xInfoPtr->num_colormap_conflicts;
	infoPtr->colormap_conflicts	= conflict;
	conflict += infoPtr->num_colormap_conflicts;
	infoPtr++;
	xInfoPtr++;
    }
    n_data = rep.n_conflicts;
    conflict = (VisualID *)(infoPtr);
    while (n_data-- > 0)
       *conflict++ = *xConflictPtr++;
    Xfree(temp_xInfo);
    Xfree(temp_conflict);
    *n_info_return = rep.n_info;
    return Success;
}
